### ISSUE TITLE
Reproducing reported numbers

### ISSUE THREAD

**Issue by thechargedneutron** [e-apostolidis/CA-SUM#1]
Hi,

Thanks for putting out the code. The instructions are also quite clear on how to run. I have a doubt regarding the steps to follow. What needs to be done after running ```sh evaluation/evaluate_exp.sh '$exp_num' '$dataset' '$eval_method'```? When I run with `$dataset` as `SumMe` and `$eval_method` as `max` I get the following result at the end

```
[Split: 0] Fscore: 42.18 [σ=0.9, epoch: 363]
[Split: 1] Fscore: 39.94 [σ=0.9, epoch: 363]
[Split: 2] Fscore: 41.74 [σ=0.9, epoch: 388]
[Split: 3] Fscore: 52.51 [σ=0.9, epoch: 359]
[Split: 4] Fscore: 47.31 [σ=0.5, epoch: 252]
Avg values :=> F1: 44.74
```

Clearly, this is much lower than the numbers reported in the paper. Do I need to go to the models and choose these epochs and then follows the steps mentioned in [this](https://github.com/e-apostolidis/CA-SUM#trained-models-and-inference) step? [EDIT]: Doing this yields the same result as above

Are there some changes that need to be done to reproduce the numbers in the paper?

[EDIT] The same difference also comes in `TVSum` training
```
[Split: 0] Fscore: 59.69  Spearman's ρ: 0.080  Kendall's τ: 0.061 [σ=0.8, epoch: 373]
[Split: 1] Fscore: 57.55  Spearman's ρ: 0.157  Kendall's τ: 0.122 [σ=0.9, epoch: 343]
[Split: 2] Fscore: 54.40  Spearman's ρ: -0.002  Kendall's τ: -0.002 [σ=0.8, epoch: 364]
[Split: 3] Fscore: 59.56  Spearman's ρ: 0.056  Kendall's τ: 0.042 [σ=0.9, epoch: 397]
[Split: 4] Fscore: 57.92  Spearman's ρ: -0.008  Kendall's τ: -0.006 [σ=0.8, epoch: 330]
====================================================================================
Avg values :=> F1: 57.82  Spearman's ρ: 0.060  Kendall's τ: 0.040
```

**Comment by tongjianglei** [comment]
I have the same problem, and I have encountered the same situation on another model PBL-SUM of the author.

**Comment by tongjianglei** [comment]
I have the same problem, and I have encountered the same situation on another model PBL-SUM of the author

**Comment by e-apostolidis** [comment]
Hi guys, and sorry for the late response.

The script "evaluate_exp.sh" should be enough for getting the final outcome about the performance of the models (both CA-SUM and PGL-SUM). The results reported in the corresponding papers, were obtained using the configurations mentioned in these works (w.r.t. the hyper-parameters of the network architectures, as well as other settings such as the seed number). All these configurations are the ones already pre-defined in our Github implementations. Moreover, as stated in our papers, the experiments of the CA-SUM method were made on a NVIDIA RTX 2080TiGPU, and the experiments of the PGL-SUM method were made on a NVIDIA TITAN Xp GPU. We also mention these details as the utilized hardware might have some impact in the evaluation outcomes, due to small changes (introduced e.g. by the applied rounding operations) in the recorded training losses, that subsequently affect the model selection process. 

Can you please let me know what GPUs do you use in your experiments? Moreover, did you try to re-produce the results reported in Table III of the PGL-SUM paper? I am asking this, because these were obtained without applying the model selection step (that might introduce changes that result in the observed performance drop) and setting batch size equal to 1.

Thanks a lot for your interest in our methods! Looking forward to your feedback!

**Comment by thechargedneutron** [comment]
I am using an NVIDIA Tesla V100 32 GB GPUs. I'm not sure if hardware implementation can significantly change the performance. I am also getting the same issue with PGL-SUM codebase. And I was not able to reproduce Table III results as well, following all the steps mentioned in both the repos. The code is extremely well-written and explained properly, thanks for that!

**Comment by tongjianglei** [comment]
Thanks for your reply. I'm also using an RTX 2080Ti GPU. it seems that the abnormal results are not caused by the hardware.

**Comment by e-apostolidis** [comment]
Hi guys. Thanks a lot for your feedback!

My initial though about the observed differences in your experiments was directed on the used GPUs, as we also did some tests with the PGL-SUM method using different GPUs (TitanXP, GTX 1080 Ti, RTX 2080 Ti) and we saw different results per GPU. To spot the issue, in each experiment we kept all the different software configurations fixed (e.g., PyTorch version, CUDA version, GPU driver version, Python version, numpy version) and only the GPU was different. To further investigate this, we run some experiments in an epoch-by-epoch manner and saw that the recorded losses were slightly different after a number of training epochs. Checking further the configurations of each GPU, we found out that the number of cores (which differs in the used GPUs) defines the size of some internally created arrays, and this size affects some calculations at the level of e.g. the fourth floating point. But such small differences are become more and more affective as the training proceeds, resulting in slightly different curves of training losses. Thus, the designed model selection criterion (that is based on these curves) might select a completely different model of the architecture, for each different GPU.

However, @tongjianglei mentioned the use of the GPU that we used in our CA-SUM work. In this case, if all the other configurations are the same, then I would expect to see the exact same results with the ones reported in our paper. So, @tongjianglei can you please confirm that you have all the other settings fixed as suggested in our repo (see section "Main dependencies")? I guess you are using the provided data splits, and not some new randomly-created ones. 

Moreover, did you try to run an experiment using the pre-trained models that we released (see "Trained models and Inference")? If yes, what is the recorded performance in this case?

Thanks again for your interest in our code!

**Comment by e-apostolidis** [comment]
Hi guys! @tongjianglei did you make another try on our code, using the experimental settings reported in our paper? @thechargedneutron did you try to use the pretrained models to see what is their performance on your GPU?

Thanks again for your interest in our method!

**Comment by JiaHN213** [comment]
Hello, it is not possible to perfectly reproduce the experimental configuration in your paper, but I experimented on GTX 1660 and TITAN RTX, and the final results differed by less than 1%, but they were very different from the results in your paper. And I look at the results using a pre-trained model, the results are the same as in your paper.

**Comment by SaBe22** [comment]
Hi,
I was also unable to reproduce the results when I trained models on TVSum splits.
The difference is particularly important for the correlation metrics (correlations around zero with my own trained models, whereas with the pre-trained models I get the same performance as in the article).
The article states that the training is done in full batch mode, so I thought the difference might come from that part of the code in the train method:
https://github.com/e-apostolidis/CA-SUM/blob/43eee5d3557e3ac06b9b8b56d1b6db03dcae16cc/model/solver.py#L87-L102

I tried moving optimizer.zero_grad() and optimizer.step() so that it is only run once at each training step, but this did not improve the results.

I don't think that all the observed differences are entirely due to hardware differences (zero correlations show that the learned scores do not match the annotations). Also, I noticed that when using a seed:
https://github.com/e-apostolidis/CA-SUM/blob/43eee5d3557e3ac06b9b8b56d1b6db03dcae16cc/model/solver.py#L26-L30
The code is missing these 2 lines to make the results hardware agnostic:
```python
torch.backends.cudnn.deterministic = True 
torch.backends.cudnn.benchmark = False
```
